Llama2 Tokenizer support

Rexwang8 opened the pull request:
This PR adds support for the following:

- Downloading of Llama2 and derivative models
- Tokenization, Detokenization using the Llama2 tokenizer
- Associated Tests

### Changes

Llama 2 follows Huggingface's nth new iteration of format with the following differences compared to previous formats.

- The tokenizer.model file, which is used to store a number of previous metadata files like the encoder.json, various configs, etc. Previously, the sentencepiece converter could be used to manually convert the .model file to the prerequisite files what we previously used. I hooked this up to the resolver.go file, and moved it to the resources folder (to prevent a circular dependency), which allows us to now download and decompose tokenizer.model files.
- The tokenizer_config.json, previously the special_tokens.json, contains a new format with some different info. The relevant info to this PR is the previous encloseEOSBOS param (Boolean) has been replaces with two, encloseEOS and encloseBOS. This allows discrete behavior which features one but not the other, which had to be accounted for. 
- Following this, I also added various encoder wrappers and an additional data folder, embedding this new data to allow support for llama2 models.
- For tests, I have 1 for downloading and 4 for the tokenizer. The 4 for the tokenizer are fairly basic "encode this corpus and compare it with the ground truth" style tests.

*Edit: The files used in the llama-tokenizer were from https://huggingface.co/KoboldAI/llama2-tokenizer

wbrown commented on gpt_bpe_test.go:
```diff
@@ -1015,6 +1052,93 @@ func TestModelDownloadPythiaSharded(t *testing.T) {
+	_, rsrcErr := resources.ResolveResources(modelId, destPathPTR,
+		resources.RESOURCE_MODEL, rsrcType, hfApiToken)
+	if rsrcErr != nil {
+		//os.RemoveAll(destPath)
```
Can we clean up and remove these `//os.RemoveAll(destPath)` lines?

wbrown commented on gpt_bpe_test.go:
```diff
@@ -1015,6 +1052,93 @@ func TestModelDownloadPythiaSharded(t *testing.T) {
+	}
+
+	// Check for pytorch_model.bin
+	modelPath := destPath + "/pytorch_model-00003-of-00003.bin"
```
This is brittle. What we should be doing is enumerating the model files and looking for the last one in the sequence.

wbrown commented on gpt_bpe.go:
```diff
@@ -102,6 +104,7 @@ const VOCAB_ID_PILE = "pile-tokenizer"
 const VOCAB_ID_CLIP = "clip-tokenizer"
 const VOCAB_ID_NERDSTASH_V1 = "nerdstash_v1-tokenizer"
 const VOCAB_ID_NERDSTASH_V2 = "nerdstash_v2-tokenizer"
+const VOCAB_ID_LLAMA_TWO = "llama_two-tokenizer"
```
why `llama_two-tokenizer` insead of `llama2-tokenizer`? How do people normally refer to it?

harubaru replied:
Apparently the Llama 2 and Llama 1 tokenizers are the same. IMO, `llama-tokenizer` would be better.

Rexwang8 replied:
i was unsure about this one, since I don't have the original copy of either, i tried to find a model from each group and tried them and they were *slightly* different, though now that I think of it, it might be differences in the downstream model. I'll rename it for now.

wbrown commented on resources/data/llama_two-tokenizer/config.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "_name_or_path": "TheBloke/Llama-2-7B-fp16",
```
Delete this specific line.

sangstar commented on gpt_bpe_test.go:
```diff
@@ -1015,6 +1052,93 @@ func TestModelDownloadPythiaSharded(t *testing.T) {
+
+	// Check for config.json
+	configPath := destPath + "/config.json"
+	if _, err := os.Stat(configPath); err == nil {
+		fmt.Println("config.json exists")
+
+	} else if errors.Is(err, os.ErrNotExist) {
+		//os.RemoveAll(destPath)
+		t.Errorf("config.json does not exist")
```
Apart from what Wes said about how this is not implemented in a brittle nor particularly extensible way, you may want to consider using a `Logger` here rather than printing to stdout.

Rexwang8 replied:
I'm not sure i want to rewrite the logging part of the entire test file, and the current method just calls for print statements, so I'm considering just leaving it for future me figure out if needed.

sangstar commented on gpt_bpe_test.go:
```diff
@@ -1015,6 +1052,93 @@ func TestModelDownloadPythiaSharded(t *testing.T) {
+		t.Errorf("config.json does not exist")
+
+	} else {
+		//os.RemoveAll(destPath)
+		t.Errorf("Error checking for config.json")
+	}
```
It would be useful for debugging if `err` itself was passed. There's probably not a lot of encounterable errors with something like this, but it's still always better to pass in `err` rather than an abstract message like this

sangstar commented on gpt_bpe.go:
```diff
@@ -160,6 +168,20 @@ func NewEncoder(vocabId string) (*GPTEncoder, error) {
+	}
+	if special, ok := (rsrcs)["tokenizer_config.json"]; ok {
+		if special.Data != nil {
+			if json.Unmarshal(*special.Data, &tokenizerSpecialConfig) != nil {
+				log.Fatal("Error unmarshalling tokenizer_config.json")
+			}
```
May be more helpful to write as

```
if special, ok := (rsrcs)["tokenizer_config.json"]; ok {
  if special.Data != nil {
          err := json.Unmarshal(*special.Data, &tokenizerSpecialConfig)
	  if err != nil {
		  log.Fatal("Error unmarshalling tokenizer_config.json: ", err)
	  }
```

As it allows you to see what the actual error was in debugging.

sangstar reviewed (commented):
Some nits

Rexwang8 commented on gpt_bpe.go:
```diff
@@ -1123,6 +1151,9 @@ func (encoder *GPTEncoder) Decode(encoded *Tokens) (text string) {
 					bs = append(bs, v...)
 				}
 			}
+			if len(bs) == 0 {
+				continue
+			}
```
ah remnant bit, there used to be a bit of code there where it would take out the start or end token if the enclose param was enabled, allowing that part of it to be fully reversable and loss-less between the original text and decode(encode()). I later found out that llama apparently doesn't do this and having a start token in the decoded string is indeed the correct output. Back before i found that out i had a line that was something like if (tokenid == start token id && enclosebos == true) then remove that safetoken. This caused some errors with weird references if it tried to progress with a bs of length 0. Since this isn't needed, will remove.

Rexwang8 commented:
Ok. I believe I touched all comments. I cleaned up the config files.
Hate to admit it, but I own HF a small apology on how i was complaining how llama enclosed the BOS token but not EOS token earlier. This was a downstream tokenizer modification. I tracked down this HF repo https://huggingface.co/KoboldAI/llama2-tokenizer  which is run by koboldAI and attempts to be a mirror of the original, which I'm now using as the ground truth. 

Updated tests to fit this. I also changed the download test from looking for an exact file to looking for the existence of the last shard (ie 3/3).

I still kept the downstream model for testing the download function, but it's not being used to provide tokenizer files anymore. 

I ran the whole project through both "gofmt" and "go fmt" and it says I'm already formatted.

Opting against adding a logger class as I think changing the tests file structure may be out of scope and having it changed for only these tests doesn't seem right.

Minor cleanup stuff here and there.

sangstar commented on gpt_bpe_test.go:
```diff
@@ -1069,7 +1071,7 @@ func TestModelDownloadLlama(t *testing.T) {
 		resources.RESOURCE_MODEL, rsrcType, hfApiToken)
 	if rsrcErr != nil {
-		//os.RemoveAll(destPath)
+		os.RemoveAll(destPath)
```
Instead of putting all these `os.RemoveAll(destPath)` everywhere for cleanup, why not just add a `defer os.RemoveAll(destPath)`?

As long as `destPath` does change after it's instantiated, I'd just suggest putting it here:

```
func TestModelDownloadLlama(t *testing.T) {
	// Pythia uses a slightly different file structure, where
	// the vocab.json and merges.txt files are stored in the
	// tokenizer.json file. We want to check if we are able to
	// download the model and extract the vocab.json and merges.txt
	modelId := "georgesung/llama2_7b_chat_uncensored"
	destPath := "./TestModelDownloadLlama"
	destPathPTR := &destPath
        defer os.RemoveAll(destPath)
```

Along with any other cleanup you do. You don't need the initial `os.RemoveAll` if cleanup is doing its job.

rtalaricw commented on resources/converter.go:
@Rexwang8 Did you write the code in this file or did we fetch this from somewhere?

Rexwang8 replied:
Wes wrote this script for converting the sentancepiece_coverter. You can find the original script under the cmd/ directory in this repo.

rtalaricw commented on gpt_bpe_test.go:
```diff
@@ -767,6 +772,40 @@ func TestUnitrimFunctionality(t *testing.T) {
+func TestLlamaTwoEncoder_Encode(t *testing.T) {
+	testString := "The fox jumped over the hare.\nThe turtle is faster than the hare."
+	llamaTokens := Llama2Encoder.Encode(&testString)
+	assert.Equal(t, llamaTokens, &Tokens{1576, 1701, 29916, 12500, 287, 975, 278, 447, 276, 29889, 13, 1576, 260, 4227, 280, 338, 8473, 1135, 278, 447, 276, 29889})
```
How did you get the "real" tokens here in assert.Equal?

rtalaricw commented on gpt_bpe_test.go:
```diff
@@ -767,6 +772,40 @@ func TestUnitrimFunctionality(t *testing.T) {
+
+func TestLlamaTwoTokenizerDecode(t *testing.T) {
+	outputString := "<s>The fox jumped over the hare.\nThe turtle is faster than the hare."
+	llamaTokens := Tokens{1, 1576, 1701, 29916, 12500, 287, 975, 278, 447, 276, 29889, 13, 1576, 260, 4227, 280, 338, 8473, 1135, 278, 447, 276, 29889}
```
Same here.

rtalaricw commented on gpt_bpe.go:
```diff
@@ -102,6 +104,7 @@ const VOCAB_ID_PILE = "pile-tokenizer"
 const VOCAB_ID_CLIP = "clip-tokenizer"
 const VOCAB_ID_NERDSTASH_V1 = "nerdstash_v1-tokenizer"
 const VOCAB_ID_NERDSTASH_V2 = "nerdstash_v2-tokenizer"
+const VOCAB_ID_LLAMA = "llama-tokenizer"
```
Is LLama-2 the same as Llama?

Rexwang8 replied:
Practically, yes. Actually, no. It's a bit of a judgement call, but I think we should separate them in a followup PR.

rtalaricw reviewed (commented):
What else is remaining here to merge this PR?

rtalaricw commented on resources/data/llama-tokenizer/tokenizer.model:
Did you use the file I sent you?

rtalaricw commented:
@Rexwang8 Can you link in the PR description where the files for the llama-tokenizer originate from?

rtalaricw reviewed (approved):
LGTM

wbrown reviewed (approved):
One slight change that I will handle in a follow up commit.
